qutip/mcsolve.py

```diff
@@ -368,964 +234,548 @@ class _MC():
-            ODE = ode(_pyRHSc)
-        ODE.set_f_params(config)
-    else:
-        ODE = ode(cy_ode_rhs)
-        ODE.set_f_params(config.h_data, config.h_ind, config.h_ptr)
-
-    ODE.set_integrator('zvode', method=opt.method, order=opt.order,
-                       atol=opt.atol, rtol=opt.rtol, nsteps=opt.nsteps,
-                       first_step=opt.first_step, min_step=opt.min_step,
-                       max_step=opt.max_step)
-    ODE.set_initial_value(config.psi0, config.tlist[0])
-    for jj in range(config.e_num):
-        expect_out[jj][0] = cy_expect_psi_csr(
-            config.e_ops_data[jj], config.e_ops_ind[jj],
-            config.e_ops_ptr[jj], config.psi0,
-            config.e_ops_isherm[jj])
-
-    for k in range(1, num_times):
-        ODE.integrate(config.tlist[k], step=0)  # integrate up to tlist[k]
-        if ODE.successful():
-            state = ODE.y / dznrm2(ODE.y)
-            for jj in range(config.e_num):
-                expect_out[jj][k] = cy_expect_psi_csr(
-                    config.e_ops_data[jj], config.e_ops_ind[jj],
-                    config.e_ops_ptr[jj], state,
-                    config.e_ops_isherm[jj])
+            return None
+
+    @property
+    def runs_states(self):
+        dims = self.psi0.dims
+        psi = np.empty((self.num_traj, len(self.tlist)), dtype=object)
+        for i in range(self.num_traj):
+            for j in range(len(self.tlist)):
+                psi[i,j] = Qobj(self._psi_out[i,j,:].reshape((-1,1)), dims=dims)
+        return psi
+
+    @property
+    def collapse(self):
+        return self._collapse
+
+    @property
+    def collapse_times(self):
+        out = []
+        for col_ in self._collapse:
+            col = list(zip(*col_))
+            col = ([] if len(col) == 0 else col[0])
+            out.append( np.array(col) )
+        return out
+        return [np.array(list(zip(*col_))[0]) for col_ in self._collapse]
+
+    @property
+    def collapse_which(self):
+        out = []
+        for col_ in self._collapse:
+            col = list(zip(*col_))
+            col = ([] if len(col) == 0 else col[1])
+            out.append( np.array(col) )
+        return out
+        return [np.array(list(zip(*col_))[1]) for col_ in self._collapse]
+
+    def get_result(self, ntraj=[]):
+        # Store results in the Result object
+        if not ntraj:
+            ntraj = [self.num_traj]
+        elif not isinstance(ntraj, list):
+            ntraj = [ntraj]
+
+        output = Result()
+        output.solver = 'mcsolve'
+        output.seeds = self.seeds
+
+        options = self.options
+        output.options = options
+
+        if options.steady_state_average:
+            output.states = self.steady_state
+        elif options.average_states:
+            output.states = self.states
+        elif options.store_states:
+            output.states = self.runs_states
+
+        if options.store_final_state:
+            output.final_state = self.final_state
```

Comment:
We could implemente the option that if  `store_final_state=True` and `average_states=True`, `Results.final_state` is a density matrix.
Instead, if `average_states=False`, the final results are the wavefunctions at the last time.

Reply:
There are ntraj final wavefunctions, returning a list of wavefunctions won't cause confusion with other solvers output which return a list of state for each time?
For now I will code it but not activate it.

Reply:
With commit "final_state average flag", `mcsolve` take `_exp=True/False`. On `True` act as you described.

Reply:
@Ericgig  : sorry, maybe what I meant was not clear.

We have three possible options which can occur: `store_states`, `store_final_state`, `average_states`.
What I think should be implemented is the following:
- if `store_states` and `store_final_state` are both `True`, the `Results.states` is an array with the states for each time step, while `Results.final_state` contains the results only for the last time.
- if `store_states=False` and `store_final_state=True`, the `Results.states` should be empty and `Results.final_state` contains the final state.

In this sense, there should be no ambiguity between `store_states` and `store_final_state`.
On top of that, the option `average_states` can be used to say if we want the wavefunctions to be averaged, or if we want the non-averaged output (i.e., one list of Qobj per trajectory). Of course, we can  either keep or remove `_exp`, whatever seems more natural.

Moreover, in this version the option `store_final_state=True` and `store_states=False`, returns a `Results.states` which is non-empty (please, find a code here https://gist.github.com/fminga/cfacb90ac4d9498fe514b6694fbee275).

Reply:
There is also the options `steady_state_average` that affect `results.states` and has priority over both `store_states` and `average_states`...
For now, I will let it like this, but we can set the default value of `_exp=True` later if better.

I hope that we use an object interface later. With that we will be able to do:
```
final_state_averaged = system.final_state
final_state_runs = system.runs_final_states
```
Thus simplifying the use of options. 

Non-empty `Results.states` is expected. By default `store_final_state=False` and `store_states=False` so there would be no outputs if no `e_ops` are given. So when no options are given, with `e_ops` you get the expectations values, otherwise you get the states. This behavior is common with other solvers (`mesolve`, `sesolve`) . If the user only want the final state, he could set `tlist=[0,end]`.

Reply:
Perfect! Thank a lot!